### DISCUSSION TITLE
ESLint/TypeScript Error Display Issues

### DISCUSSION THREAD

**Discussion by shellscape** [zed-industries/zed#33545] [Feature Requests]
### Summary

The display of both popup and inline errors for ESLint and TypeScript needs some fine tuning.

### Overview of undesirable behavior

Starting with a TypeScript node project with a valid `tsconfig.json` and `.eslintrc`

We'll start with an innocuous code block:

<img width="494" alt="Screenshot 2023-05-15 at 9 34 16 AM" src="https://github.com/zed-industries/community/assets/76371/621790fb-8fdc-4f5f-9571-bfaa4a6ef7df">

And then add a broken/incomplete `if` statement:

<img width="476" alt="Screenshot 2023-05-15 at 9 34 31 AM" src="https://github.com/zed-industries/community/assets/76371/93b35ad7-7da9-4dd9-b707-62122e48417e">

We can see above that the type info + ESLint error info popups are disjointed, and take up a lot of visual space. This is really uncomfortable to look at and I much prefer the unified popup experience in Atom/VSCodium. That's the first area of polish I'd like to see.

<img width="438" alt="Screenshot 2023-05-15 at 9 35 37 AM" src="https://github.com/zed-industries/community/assets/76371/49d42d05-4f67-4c78-8436-36e1d604b6f8">

Next, we'll click on the statusbar indicator for the error, which results in:

<img width="434" alt="Screenshot 2023-05-15 at 9 35 47 AM" src="https://github.com/zed-industries/community/assets/76371/39d1d63c-2f47-42c4-8903-7c08435c07a5">

Now we've got two popups, and an inline/under-current-line error indicator. This is non-optimal.

<img width="467" alt="Screenshot 2023-05-15 at 9 35 57 AM" src="https://github.com/zed-industries/community/assets/76371/1c7e80ad-d8fc-45b6-9c31-f5049577ff99">

Then we'll highlight and delete that block in question:

<img width="467" alt="Screenshot 2023-05-15 at 9 35 57 AM" src="https://github.com/zed-industries/community/assets/76371/6891947b-9a43-4a74-a7ae-76531efdacf0">

Which then results in:
<img width="433" alt="Screenshot 2023-05-15 at 9 36 03 AM" src="https://github.com/zed-industries/community/assets/76371/0b7fb2ca-1fe8-4c9e-853f-77e7d2699616">

Notice how the inline error has stayed in the editor. It's not editable nor removable, and I have to close the file and reopen to get rid of it. Sometimes that text stays red, other times it doesn't. It also appears to jump around a lot as to where its positioned, following the same procedure at the same positions as shown above.

So, there's a few issues with ESLint support yet. 

### Environment

Zed: v0.85.4 (stable)
OS: macOS 13.3.1
Memory: 16 GiB
Architecture: aarch64

**Comment by JosephTLyons** [comment]
Hey @shellscape, just wanted to mention that the inline errors and warnings can be dismissed by hitting escape.  We will look for a way to make dismissing these more discoverable.  As for having multiple popup windows, we'll have to wait for someone on the design team to weigh in.

**Comment by shellscape** [comment]
Thanks for the info. I think you still may have a big with the diag message persisting after the offending code has been removed.

**Comment by JosephTLyons** [comment]
Yeah, the diagnostic count being out of sync with the language server has been a huge pain - it's constantly being discussed in meetings.  We have some ideas for it, but its a tough problem.

**Comment by evelant** [comment]
Perhaps a related issue -- eslint will recommend autofix in some cases but running the suggested autofix appears to be missing from the ctrl+. actions menu.

**Comment by shellscape** [comment]
I'd also like to add additional nuance to this: Zed will display the source (eslint) and the message of the error, but not the ESLint rule name. VS Code's display is unfortunately vastly superior here, as it provides all of the mentioned info plus suggestions on how to remedy.

**Comment by shinebayar-g** [comment]
How do you get eslint working? It's not even showing up on my editor.

**Comment by felixaa** [comment]
> How do you get eslint working? It's not even showing up on my editor.

Same here, cannot get eslint warnings/errors to display. Only typescript errors shows up
<img width="302" alt="image" src="https://github.com/zed-industries/zed/assets/5445088/35512749-cc69-40bd-b2d2-01ec74af5e1c">

**Comment by nadavmoreno** [comment]
using 0.125.3
also can't see eslint warning/errors to display

**Comment by Spoutnik97** [comment]
I think it is related to this issue: I can't move my mouse hover the popup; I can't see the whole message, so I can't debug it ...

Watch the video : 

https://github.com/zed-industries/zed/assets/30442116/e9977ed9-1d5d-442c-b72f-ae0a08028a13

**Comment by Spoutnik97** [comment]
But sometimes it works

**Comment by labithiotis** [comment]
TS errors are rendered on a single-line making it very hard to read errors.

![image](https://github.com/user-attachments/assets/04514684-e24c-4e20-afdf-52daa04a0c6d)

OS: M1 Mac
Zed: 0.147.2

**Comment by dobernike** [comment]
Falsy parsing error on eslint in TSX file:

![image](https://github.com/user-attachments/assets/7f5e33df-8232-429c-b925-273f6efb90ae)

![image](https://github.com/user-attachments/assets/5792ec6a-26d0-489e-b668-c02ad8854857)

**Comment by stepankuzmin** [comment]
There are also TypeScript errors on commented out blocks of code

<img width="411" height="211" alt="Screenshot 2026-06-03 at 11 09 40" src="https://github.com/user-attachments/assets/7e72250e-a534-4c53-8eff-51a90a88e144" />
